akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala

```diff
@@ -41,119 +39,57 @@ final private[stream] class OutputStreamSourceStage(writeTimeout: FiniteDuration
 
     require(maxBuffer > 0, "Buffer size must be greater than 0")
 
-    val dataQueue = new LinkedBlockingQueue[ByteString](maxBuffer)
+    // Semaphore counting the number of elements we are ready to accept,
+    // which is the demand plus the size of the buffer.
+    val semaphore = new Semaphore(maxBuffer, /* fair =*/ true)
+
     val downstreamStatus = new AtomicReference[DownstreamStatus](Ok)
 
     final class OutputStreamSourceLogic extends GraphStageLogic(shape) {
 
-      var flush: Option[Promise[Unit]] = None
-      var close: Option[Promise[Unit]] = None
-
-      private var dispatcher: ExecutionContext = null // set in preStart
-      private val blockingThreadRef: AtomicReference[Thread] = new AtomicReference() // for postStop interrupt
-
-      private val downstreamCallback: AsyncCallback[Try[ByteString]] =
-        getAsyncCallback {
-          case Success(elem) ⇒ onPush(elem)
-          case Failure(ex)   ⇒ failStage(ex)
-        }
-
-      private val upstreamCallback: AsyncCallback[(AdapterToStageMessage, Promise[Unit])] =
+      private val upstreamCallback: AsyncCallback[AdapterToStageMessage] =
         getAsyncCallback(onAsyncMessage)
 
-      def wakeUp(msg: AdapterToStageMessage): Future[Unit] = {
-        val p = Promise[Unit]()
-        upstreamCallback.invoke((msg, p))
-        p.future
-      }
+      def wakeUp(msg: AdapterToStageMessage): Unit =
+        upstreamCallback.invoke(msg)
```

Comment:
That's not waking anything up anymore though? Maybe just skip this and do ETA expansion when we pass it to the adapter?